src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java

```diff
@@ -368,11 +429,50 @@ private void setGlueSchemaIntoUserRecordIfAny(UserRecord userRecord, Message<?>
 	}
 
 	private CompletableFuture<UserRecordResponse> handleUserRecord(UserRecord userRecord) {
+		if (this.maxInFlightRecords != -1) {
+			waitForCapacityInUserRecordsBuffer();
+		}
+
 		ListenableFuture<UserRecordResult> recordResult = this.kinesisProducer.addUserRecord(userRecord);
 		return listenableFutureToCompletableFuture(recordResult)
 				.thenApply(UserRecordResponse::new);
 	}
 
+	private void waitForCapacityInUserRecordsBuffer() {
```

Comment:
Its a good thought, sharing my opinions from references with the spring cloud stream kinesis binder. What you are suggesting,  can be very helpful if any consumer is using the spring integration directly rather than transitively. 

1. Any consumer using the spring cloud stream kinesis binder - would require ```KinesisProducer``` exposed as a bean that can be used.  The binder internally constructs ```KinesisProducer```  rather than referencing a bean. 
2.  This can still be baked into the ```KPLMessageHandler``` given, all the consumers of the handler would be replicating the same logic when needed.  As a consumer, it can be a great thought, if the internals are intelligent. 
3. For accomodating the backpressure if we need to use the Resilience Advices - we dont need this change. 

Please suggest the next steps.

Reply:
Well, technically Spring Integration is fully used by Spring Cloud Stream transitively. 
In most cased we indeed don't have direct access to those `MessageHandler` instances used in binders.
However we still can improve those binders to expose some hooks or configuration properties.
In the end it won't matter of end-user if those properties are mapped to the `KPLMessageHandler` or some `RateLimiterRequestHandlerAdvice` and `RequestHandlerRetryAdvice` combination.

I'm not sure why would one need `KinesisProducer` instance from binder, if the logic is going to be covered by respective exposed configuration properties for the binder producer.

I feel that something like `maxInFlightRecords` and some `KplBackPressureException` should be enough for this component. The rest of the logic could be handled by the `RequestHandlerRetryAdvice`.

Reply:
Fixed, as discussed. 
Changed the handling to a simpler condition. So that the users can use the Spring advice as needed.